src/main/java/frc/robot/subsystems/CoralManipulator.java

```diff
@@ -0,0 +1,143 @@
+
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.networktables.DoublePublisher;
+import edu.wpi.first.networktables.BooleanPublisher;
+import edu.wpi.first.units.measure.Angle;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.WaitUntilCommand;
+import frc.robot.Constants;
+import frc.util.control.nt.PIDTuning;
+import frc.util.control.nt.TunableNumber;
+
+import java.util.function.BooleanSupplier;
+
+public class CoralManipulator extends SubsystemBase {
+    private final TalonFX pivotMotor;
+    private final TalonFX rollerMotor;
+    private final DigitalInput beambreak;
+    private final MotionMagicVoltage motionMagic;
+    
+    private final TunableNumber kP;
+    private final TunableNumber kI;
+    private final TunableNumber kD;
+    private final PIDTuning pidTuning;
+
+    private final NetworkTable manipulatorTable;
+    private final DoublePublisher anglePub;
+    private final BooleanPublisher beambreakPub;
+    private final DoublePublisher currentPub;
+    
+    private boolean openLoop;
+    private Angle targetAngle;
+
+
+    public CoralManipulator() {
+        pivotMotor = Constants.CoralManipulator.PIVOT_CONFIG.createTalonForMotionMagic();
+        rollerMotor = Constants.CoralManipulator.ROLLER_CONFIG.createTalon();
+        beambreak = new DigitalInput(Constants.Ports.BEAM_BREAK_PORT);
+  
+        motionMagic = new MotionMagicVoltage(0);
+        
+        kP = new TunableNumber("Manipulator/PID/kP", Constants.CoralManipulator.kP, Constants.TUNING_MODE);
+        kI = new TunableNumber("Manipulator/PID/kI", Constants.CoralManipulator.kI, Constants.TUNING_MODE);
+        kD = new TunableNumber("Manipulator/PID/kD", Constants.CoralManipulator.kD, Constants.TUNING_MODE);
+        pidTuning = Constants.CoralManipulator.PIVOT_CONFIG.genPIDTuning("Manipulator Pivot", Constants.TUNING_MODE);
+
+        // configure pivot motor
+        var config = new TalonFXConfiguration();
+        config.Slot0.kP = Constants.CoralManipulator.kP;
+        config.Slot0.kI = Constants.CoralManipulator.kI;
+        config.Slot0.kD = Constants.CoralManipulator.kD;
+        
+        config.SoftwareLimitSwitch.ForwardSoftLimitEnable = true;
+        config.SoftwareLimitSwitch.ForwardSoftLimitThreshold = Constants.CoralManipulator.MAX_ANGLE.in(Radians);
+        config.SoftwareLimitSwitch.ReverseSoftLimitEnable = true;
+        config.SoftwareLimitSwitch.ReverseSoftLimitThreshold = Constants.CoralManipulator.MIN_ANGLE.in(Radians);
+        
+        // current and voltage limits
+        config.CurrentLimits.SupplyCurrentLimit = Constants.CoralManipulator.CURRENT_LIMIT.in(Amps);
+        config.CurrentLimits.SupplyCurrentLimitEnable = true;
+        
+        pivotMotor.getConfigurator().apply(config);
```

Comment:
i think we should refactor this extra configuration into motor config since we are doing it in the elevator and here. Also, we probably don't need motion magic for this mechanism